Closes issue #60 and #57

karthikp-ubc opened the pull request:
Added injection capability for both src and dest registers - allreg option. Need to update the documentation in the wiki.

abrahamchan commented on llvm_passes/core/Controller.cpp:
```diff
@@ -73,8 +73,8 @@ static cl::opt<FIRegSelMethod> firegselmethod(
     cl::values(clEnumVal(dstreg, "Inject into destination register"),
-               clEnumVal(allsrcreg,
-                         "Inject randomly into one of all source registers"),
+               clEnumVal(allreg, "Injection randomly into either destination register or one of the source registers"),
```
Perhaps this is being pedantic, but could you please change this from 'injection' into 'inject'?

abrahamchan commented on llvm_passes/core/FaultInjectionPass.cpp:
```diff
@@ -93,6 +93,15 @@ void FaultInjectionPass::insertInjectionFuncCall(
+      // The return type is not a valid return type, and should hence be ignored
+      // This is to deal with types such as Metadata that are not valid return types
+      // Or if the instruction is an intrinsic one (starts with 'llvm_', ignore it
+      if (! FunctionType::isValidReturnType(returntype) 
```
Could you please place the ! closer to the expression? `!FunctionType::isValidReturnType(returntype)`

abrahamchan commented on llvm_passes/core/FaultInjectionPass.cpp:
```diff
@@ -93,6 +93,15 @@ void FaultInjectionPass::insertInjectionFuncCall(
+      // This is to deal with types such as Metadata that are not valid return types
+      // Or if the instruction is an intrinsic one (starts with 'llvm_', ignore it
+      if (! FunctionType::isValidReturnType(returntype) 
+	 || isa<IntrinsicInst>(fi_inst) ) 
```
The usual LLVM coding style is to place the || at the end of the previous line.

abrahamchan commented on sample_programs/cpp_sample_programs/factorial/input.yaml:
```diff
@@ -7,13 +7,9 @@ compileOption:
 
     regSelMethod: regloc
-    regloc: dstreg
+    regloc: allreg
```
Should we also add this option `allreg` to the master YAML file, under `docs/input_masterlist.yaml` and `docs/input_masterlist_ml.yaml` ?

karthikp-ubc replied:
Yes, we should - I didn't know how to do this via a PR though. The easiest way would be to directly edit the Wiki page

abrahamchan reviewed (changes requested):
Thanks for submitting this PR. Just left some minor comments.

karthikp-ubc commented:
I've addressed the above comments. I also updated the input.yaml files for a bunch of different program to test the allreg functionality, but I'll merge it in a separate PR after this one has been merged. Let me know if there're any other comments.

abrahamchan commented on llvm_passes/core/FaultInjectionPass.cpp:
```diff
@@ -93,6 +93,15 @@ void FaultInjectionPass::insertInjectionFuncCall(
+      // This is to deal with types such as Metadata that are not valid return types
+      // Or if the instruction is an intrinsic one (starts with 'llvm_', ignore it
+      if ( !FunctionType::isValidReturnType(returntype) ||
+		 isa<IntrinsicInst>(fi_inst) ) 
```
Please align the `isa<IntrinsicInst>(fi_inst)` with the same indent as `!FunctionType`. Otherwise, everything else looks good.

karthikp-ubc commented:
Just addressed the comment. I'm going to merge this PR now. Thanks.
